crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -130,6 +160,15 @@ impl CheckpointExecutor {
             .checkpoint_exec_epoch
             .set(epoch_store.epoch() as i64);
 
+        // check if we want to run this epoch based on RunWithRange condition value
+        if run_with_range.matches_epoch(epoch_store.epoch()) {
+            info!(
+                "RunWithRange condition satisfied at epoch {:?}",
+                epoch_store.epoch()
+            );
+            return StopReason::RunWithRangeCondition;
+        }
+
```

Comment:
To be clear, `Epoch(5, true)` would mean execute epoch 5, then reconfigure, so we end execution in the beginning of epoch 6 (like snapshots), and `Epoch(5, false)` would mean execute to end of epoch 5 but don't reconfigure.

Reply:
if we instead checked `if epoch_store.epoch() > run_with_range` and always reconfigured, would that work? Or are there use cases for not reconfiguring? I'm fine with adding the WithReconfiguration flag otherwise.

Reply:
I ran the run-with-range checkpoint mode on a sui-node over the weekend. It looks like specifying the last checkpoint of an epoch does exit before a reconfigure occurs. 

logs:
```
Jan 13 01:07:07 gorpig systemd[1]: sui-node.service: Consumed 4d 3h 3min 7.238s CPU time.
Jan 13 01:07:07 gorpig systemd[1]: sui-node.service: Deactivated successfully.
Jan 13 01:07:04 gorpig sui-node[3632024]: {"timestamp":"2024-01-13T01:07:04.921926Z","level":"INFO","fields":{"message":"Encountered request error (error sending request for url (https://s3.us-west-2.amazonaws.com/mysten-testnet-archives/epoch_9/685784.sum): error try>
Jan 13 01:07:04 gorpig sui-node[3632024]: {"timestamp":"2024-01-13T01:07:04.921724Z","level":"INFO","fields":{"message":"Encountered request error (error sending request for url (https://s3.us-west-2.amazonaws.com/mysten-testnet-archives/epoch_9/683831.sum): error try>
Jan 13 01:07:04 gorpig sui-node[3632024]: {"timestamp":"2024-01-13T01:07:04.921522Z","level":"INFO","fields":{"message":"Encountered request error (error sending request for url (https://s3.us-west-2.amazonaws.com/mysten-testnet-archives/epoch_9/681880.sum): error try>
Jan 13 01:07:04 gorpig sui-node[3632024]: {"timestamp":"2024-01-13T01:07:04.921314Z","level":"INFO","fields":{"message":"Encountered request error (error sending request for url (https://s3.us-west-2.amazonaws.com/mysten-testnet-archives/epoch_9/679966.sum): error try>
Jan 13 01:07:04 gorpig sui-node[3632024]: {"timestamp":"2024-01-13T01:07:04.921065Z","level":"INFO","fields":{"message":"Encountered request error (error sending request for url (https://s3.us-west-2.amazonaws.com/mysten-testnet-archives/epoch_9/678011.sum): error try>
Jan 13 01:06:40 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:06:10 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporter otlp encountered the following error(s): the grpc server returns error (The service is currently unavailable): , detailed error message: error trying to connect: tcp connect error: >
Jan 13 01:05:59 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:05:29 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:04:59 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:04:29 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:03:59 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporter otlp encountered the following error(s): the grpc server returns error (The service is currently unavailable): , detailed error message: error trying to connect: tcp connect error: >
Jan 13 01:03:48 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:03:18 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:02:48 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:02:18 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:01:48 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporter otlp encountered the following error(s): the grpc server returns error (The service is currently unavailable): , detailed error message: error trying to connect: tcp connect error: >
Jan 13 01:01:37 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:01:07 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:00:37 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:00:07 gorpig sui-node[3632024]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 13 01:00:07 gorpig sui-node[3632024]: {"timestamp":"2024-01-13T01:00:07.490933Z","level":"INFO","fields":{"message":"Future narwhal/network/src/connectivity.rs:ConnectionMonitor completed"},"target":"narwhal_network::connectivity","filename":"narwhal/network/src/c>
Jan 13 01:00:07 gorpig sui-node[3632024]: {"timestamp":"2024-01-13T01:00:07.464844Z","level":"INFO","fields":{"message":"RunWithRange condition satisifed after checkpoint sequence number 176295"},"target":"sui_core::checkpoints::checkpoint_executor","filename":"crates
```

for contrast this is a sui-node run using epoch 1 as the delimiter. Here you can see it does reconfigure and stop at the nest epoch.
```
Jan 11 23:25:30 gorpig sui-node[3578737]: {"timestamp":"2024-01-11T23:25:30.474624Z","level":"DEBUG","fields":{"message":"Epoch terminated - waiting for pending tasks to complete"},"target":"sui_core::authority::authority_per_epoch_store","filename":"crates/sui-core/s>
Jan 11 23:25:30 gorpig sui-node[3578737]: {"timestamp":"2024-01-11T23:25:30.474678Z","level":"DEBUG","fields":{"message":"All pending epoch tasks completed"},"target":"sui_core::authority::authority_per_epoch_store","filename":"crates/sui-core/src/authority/authority_>
Jan 11 23:25:30 gorpig sui-node[3578737]: {"timestamp":"2024-01-11T23:25:30.494376Z","level":"INFO","fields":{"message":"Node State has been reconfigured","next_epoch":2},"target":"sui_node","filename":"crates/sui-node/src/lib.rs","line_number":1623}
Jan 11 23:25:30 gorpig sui-node[3578737]: {"timestamp":"2024-01-11T23:25:30.494702Z","level":"INFO","fields":{"message":"Reconfiguration finished"},"target":"sui_node","filename":"crates/sui-node/src/lib.rs","line_number":1576}
Jan 11 23:25:30 gorpig sui-node[3578737]: {"timestamp":"2024-01-11T23:25:30.640096Z","level":"INFO","fields":{"message":"RunWithRange condition satisfied at Some(Epoch(1)), run_epoch=2"},"target":"sui_core::checkpoints::checkpoint_executor","filename":"crates/sui-core>
Jan 11 23:25:30 gorpig sui-node[3578737]: {"timestamp":"2024-01-11T23:25:30.644598Z","level":"INFO","fields":{"message":"Future narwhal/network/src/connectivity.rs:ConnectionMonitor completed"},"target":"narwhal_network::connectivity","filename":"narwhal/network/src/c>
Jan 11 23:25:35 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:26:05 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:26:35 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:27:05 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:27:16 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporter otlp encountered the following error(s): the grpc server returns error (The service is currently unavailable): , detailed error message: error trying to connect: tcp connect error: >
Jan 11 23:27:46 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:28:16 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:28:46 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:29:16 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:29:27 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporter otlp encountered the following error(s): the grpc server returns error (The service is currently unavailable): , detailed error message: error trying to connect: tcp connect error: >
Jan 11 23:29:57 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:30:27 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:30:57 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:31:27 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:31:38 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporter otlp encountered the following error(s): the grpc server returns error (The service is currently unavailable): , detailed error message: error trying to connect: tcp connect error: >
Jan 11 23:32:08 gorpig sui-node[3578737]: OpenTelemetry trace error occurred. Exporting timed out after 30 seconds
Jan 11 23:32:12 gorpig sui-node[3578737]: {"timestamp":"2024-01-11T23:32:12.567520Z","level":"INFO","fields":{"message":"Encountered request error (error sending request for url (https://s3.us-west-2.amazonaws.com/mysten-testnet-archives/epoch_5/340426.sum): error try>
Jan 11 23:32:12 gorpig sui-node[3578737]: {"timestamp":"2024-01-11T23:32:12.567783Z","level":"INFO","fields":{"message":"Encountered request error (error sending request for url (https://s3.us-west-2.amazonaws.com/mysten-testnet-archives/epoch_5/342065.sum): error try>
Jan 11 23:32:12 gorpig sui-node[3578737]: {"timestamp":"2024-01-11T23:32:12.567990Z","level":"INFO","fields":{"message":"Encountered request error (error sending request for url (https://s3.us-west-2.amazonaws.com/mysten-testnet-archives/epoch_5/343738.sum): error try>
Jan 11 23:32:12 gorpig sui-node[3578737]: {"timestamp":"2024-01-11T23:32:12.568197Z","level":"INFO","fields":{"message":"Encountered request error (error sending request for url (https://s3.us-west-2.amazonaws.com/mysten-testnet-archives/epoch_5/345297.sum): error try>
Jan 11 23:32:12 gorpig sui-node[3578737]: {"timestamp":"2024-01-11T23:32:12.568444Z","level":"INFO","fields":{"message":"Encountered request error (error sending request for url (https://s3.us-west-2.amazonaws.com/mysten-testnet-archives/epoch_5/346843.sum): error try>
Jan 11 23:32:13 gorpig systemd[1]: sui-node.service: Deactivated successfully.
Jan 11 23:32:13 gorpig systemd[1]: sui-node.service: Consumed 1d 14h 6min 59.571s CPU time.
Jan 11 23:32:14 gorpig systemd[1]: sui-node.service: Scheduled restart job, restart counter is at 1.
Jan 11 23:32:14 gorpig systemd[1]: Stopped sui-node.service.
Jan 11 23:32:14 gorpig systemd[1]: sui-node.service: Consumed 1d 14h 6min 59.571s CPU time.
Jan 11 23:32:14 gorpig systemd[1]: Started sui-node.service.
```